### ISSUE TITLE
[Feature] - Support for render-to-string and partial-hydration

### ISSUE THREAD

**Issue by matthewharwood** [schell/mogwai#30]
Preface:
I think Mogwai has a huge potential and thank you for making it!  

The `cargo generate` .wasm bundle is the smallest I've seen uncompressed 48kb to 18kb gz and prob 25% smaller Brotli! Not too shabby, and therefore, you should promote these number in the github `README.md`:
![image](https://user-images.githubusercontent.com/1505266/85230919-0a7c0900-b3a8-11ea-8bfa-ca2fba373bdb.png)

And while I haven't personally battle tested, it would seem your benchmarks are superior to all of the alternatives:

![image](https://user-images.githubusercontent.com/1505266/85230712-dbb16300-b3a6-11ea-89f5-135bfe929bb1.png)

I think one more thing to promote is the benefit that wasm is pre-parsed/compiled which makes the cost of JavaScript far smaller:
![image](https://user-images.githubusercontent.com/1505266/85231018-c806fc00-b3a8-11ea-8395-24806c419220.png)

**Background**

I more frequently make (over-engineer) large(100k-1m pages) marketing sites. When building these sites I have always some key foci: Developer experience, speed & performance, SEO/a11y , and rich user experience.

I'm personally looking for a no-compromise framework; one that affords the opportunity to build static sites or a robust application. 

**Feature Request**

There are 3 features I'd like to see from Mogwai
1. render-to-string
2. Partial/selective hydration:
3. code-splitting (I don't think it is possible in wasm, yet)

**Render-to-String**

In something like preact it would look like this:

```
const html = require('htm');
const renderToString = require('preact-render-to-string');

const App = (props) => html`
  <div class="app">
    <h1>This is an app</h1>
    <p>Current server time: ${new Date + ''}</p>
  </div>
`;
const doc = renderToString(App);
```

Output
```
  <div class="app">
    <h1>This is an app</h1>
    <p>Current server time: Sun Jun 21 2020 10:55:20 GMT-0700 (Pacific Daylight Time) </p>
  </div>
```
The power in this is we could write Mogwai and generate a pure js static website.

That said, 100% JavaScript free website can be boring.  We could make make a runtime that can rehydrate the prerendered html from the render-to-string fn above:

[Link](https://developers.google.com/web/updates/2019/02/rendering-on-the-web#rehydration-issues)
> A Rehydration Problem: One App for the Price of Two
Rehydration issues can often be worse than delayed interactivity due to JS. In order for the client-side JavaScript to be able to accurately “pick up” where the server left off without having to re-request all of the data the server used to render its HTML, current SSR solutions generally serialize the response from a UI’s data dependencies into the document as script tags. The resulting HTML document contains a high level of duplication:

![image](https://user-images.githubusercontent.com/1505266/85231496-47e29580-b3ac-11ea-82bd-51350798ce6f.png)

This full hydration step is pretty status quo today.  It gets a bit better with routing and this is 100% ok and personally I'm 100% ok with this approach.  

But if you're curious, theirs a new approach called selective or partial hydration.  The key here is to select the components that may need hydration and essentially wrap them. I've found a lot of benefit doing this hydration with web components:

```
import {RunningHeader} from '../running-header';

export function autoHydrate(Component, name) {
 <hydrate-root name=${name} />
    <${Component} ...${props} />
    <script
        type="text/hydration"
        dangerouslySetInnerHTML=${{
      __html: JSON.stringify({props})
    }}
    />

const HydratedRunningHeader = autoHydrate(RunningHeader, {name: 'hello});
```

```
import {HydratedRunningHeader} from '../components/running_header';
import {h, hydrate} from 'preact';


class HydrateRoot extends HTMLElement {
  constructor() {
    super();
  }
  connectedCallback() {
    const childNodes = [];
    let $end = this;
    let data = {};
    
    while (($end = $end.nextSibling)) {
      if (
          $end.nodeName === "SCRIPT" &&
          $end.getAttribute("type") === "text/hydration"
      ) {
        try {
          data = JSON.parse($end.textContent) || {};
        } catch (e) {}
        break;
      }
      childNodes.push($end);
    }

    this.root = {
      childNodes,
      appendChild: c => this.parentNode.insertBefore(c, $end)
    };
    hydrate(h(HydratedRunningHeader, data.props), this.root);
  }
}
export {HydrateRoot};
```

If we merge these two ideas we could take the original idea and pass in props:
```
const html = require('htm');
const renderToString = require('preact-render-to-string');

const App = (props) => html`
  <div class="app">
    <h1>This is an ${props.name}</h1>
    <p>Current server time: ${new Date + ''}</p>
  </div>
`;
const HydratedApp = autoHydrate(App , {name: 'HELLO WORLD'});
const doc = renderToString(HydratedApp);
```

Output:
```
  <hydrate-root name='App' />
  <div class="app">
    <h1>This is an HELLO WORLD</h1>
    <p>Current server time: Sun Jun 21 2020 10:55:20 GMT-0700 (Pacific Daylight Time) </p>
  </div>
 <script type="text/hydration">
   {props: {name: 'HELLO WORLD'}}
 </script>
```

When the hydrated-root bootstraps it will read the dom until it hits the script tag, it will JSON.parse the props and `hydrate` them back into the component.  

In this example, with this approach, you can selectively or paritally bundle only the interactive components of an application/page and therefore, make the **absolute smallest JavaScript bundle possible**.

Ok this was a huge post but would love to hear back from you... in the mean time i'mma give mogwai a shot lol.

**Comment by d4h0** [comment]
> 1) render-to-string
> 2) Partial/selective hydration:
> 3) code-splitting (I don't think it is possible in wasm, yet)

I, as well, think these features are very important for many types of websites (especially, content-heavy sites like online shops, blogs, marketing pages, etc.)

But also for most other websites. Because, without server-side rendering, most bots won't work.

Googles Search Bot executes pages in a real browser (however there seem to be a few things to consider), but most other bots do not run a real browser to execute JavaScript and WebAssembly. 

That means, for example, if someone shares a page powered by mogwai on Facebook, Facebook can't extract metadata from the actual page that is linked to.

The same problem you will get with most other bots. And this is not the only kind of problem. So I think SSR is very important.

I just don't understand why someone would want to add JavaScript. I'm mostly interested in WebAssembly so I can eliminate JavaScript as much as possible... ;)

**Comment by schell** [comment]
This is an epic ticket, I'm flattered! You both make good points. I simply haven't thought about SSR that much. I think it's a good thing to have though I haven't done much research (@matthewharwood thank you for the head start). It will take a bit of time for me to digest the state of the art and decide on the next mogwai milestone. In the meantime please let me know if there are any clarifications I can make to the docs or examples :)

**Comment by matthewharwood** [comment]
Theres a really good conversation about this topic also on SolidJS github issues: https://github.com/ryansolid/solid/issues/109#issuecomment-649061739

**Comment by matthewharwood** [comment]
I just found this project that has an SSR implementation in rust/wasm...

https://github.com/chinedufn/percy

**Comment by schell** [comment]
Thanks for the heads up @matthewharwood. Although I've been rather silent here the past few weeks, I have been thinking about this and gathering my thoughts in a doc elsewhere. I think I have a strategy and will add it here for your thoughts and then I'll add it to the roadmap as soon as it's viable.

**Comment by matthewharwood** [comment]
@schell no problem at all! this is great to hear! I've actually hit the books for rust proper... My skills are just a bit too weak with general purpose statically typed languages. I look forward to coming back to Mogwai when I've come out of this rabbit hole!

**Comment by schell** [comment]
`mogwai` uses `web-sys` and `wasm_bindgen` to accomplish DOM creation and dynamic updates. 
Mogwai's main type `Gizmo<T>` has a type variable `T` meant to be filled by `web-sys` types
that represent DOM nodes. Since `web-sys`'s operations rely on javascript->wasm interop, any server side (ie rust only) pre-rendering should not happen through the normal browser-bound mogwai codepath. Of course using `web-sys` in 
this way would compile fine - but attempting to build the DOM and run a mogwai Gizmo when compiled to native code would cause a panic because the DOM API is not available outside of the browser. Not all is lost though, as this actually points us in the direction of
a sane implementation so long as we don't get carried away. 

`mogwai` is meant to be a lightning fast frontent library, not a feature-rich web development framework. There
are already plenty of heavy frameworks (I'm looking at Yew, kid). To this end I don't want to push `mogwai` into 
becoming an "isomorphic" framework (ie one that can run the same code on both server and browser). Instead I would like 
`mogwai` to have a compelling SSR story that ties in well with its other goals:

* be fast
* be small
* be explicit 

Additionally I'd like to support SSR without requiring more effort from library users. 

## A solution

At the very least we need to add a function like

```rust
fn gizmo_to_string(gizmo:Gizmo) -> String;
```

which could be used to serialize a gizmo to a string. But of course this multiplies the amount of effort needed to define a `mogwai` component that can be serialized server-side. The library user now needs to define the `view` function and the `gizmo_to_string` function. Another alternative would be to define components using a builder (which was the case for `mogwai` v0.1:

```rust
fn view(...) -> GizmoBuilder<T>;
```

and then use this `GizmoBuilder` to build the DOM in the browser and to build the `String` on the server. However, this approach negatively affects DOM building performance, which is a major tenet of `mogwai` in the first place. The *idea* of a builder is a step in the right direction in my opinion, as it clarifies what we're really looking for here - _one definition of the view that can be used on the server and the browser_. The problem isn't trivial but if we enumerate the requirements we'll see that one of rust's tools is a good fit for the job:

* define a component once with no added library user effort
* build and run that component in the browser with no added performance cost 
* build and serialize that component on a server
* only pay for what you use, ie - if you don't want SSR you should not have to change your code and there should be no compile-time or performance cost

tl;dr Define a component once and use that definition in multiple contexts without an intermediate abstraction. 

This sounds like a job for procedural macros! If we had a macro DSL for defining components we could transform this definition:

```rust
ssr_component!(
  <div class="my_class">...</div>
)
```

into

```rust 
fn view(...) -> Gizmo<...> {
    div()
      .class("my_class")
      ...
}

fn to_string(...) -> Gizmo<...> {
    r#"<div class="my_class">...</div>"#
}
```

In general I've been apprehensive about writing a JSX macro but it looks like it serves a nice use case here, and using a procedural macro opens up some interesting possibilities for separating the view into html(ish) templates.

## unsolved
This doesn't solve hydration but I think a solution for that will shake out after playing around with this macro stuff.

**Comment by schell** [comment]
`percy` has the best jsx/html macros i've seen so far, so I'll probably lift a lot of inspiration from that project.

**Comment by matthewharwood** [comment]
Hey sorry for the lateness in reply... Been a crazy month for me.  I love this idea and thanks for the hard work.

Could "build speeds" be added as a core consideration too.  These render-to-string should be able to build in parallel. Imagine rendering a 1, 100, to a million components at once. I know that these are big concerns in the JS 

> This doesn't solve hydration but I think a solution for that will shake out after playing around with this macro stuff

It's still a wip but I'm doing a presentation on partial hydration and jam stack.  I love to walk you through it and see what ideas we can apply.   Alot of this is uncharted territory and I have this hunch that this could lead to a big. If so, We could go by email probably easier to schedule. matthhar12@gmail.com

**Comment by matthewharwood** [comment]
Also this might help with the macro https://github.com/Pauan/rust-dominator/issues/33

**Comment by schell** [comment]
You can reach me at efsubenovex@gmail.com.

**Comment by schell** [comment]
I have a working solution to render-to-string in PR #34. It uses conditional compilation to include some extra fields on mogwai's main node type. This enables the traversal of attributes and children even when we can't access a `web_sys::Node`, eg any time a node is compiled server-side.

This PR changes the node building API significantly and I think it's for the better. The API has fewer calls and they've been somewhat normalized. I've removed the specific `id`, `class` etc attribute funcions and instead there is one `attribute` function and one `style` function that take a polymorphic angument: an immediate value `T` or a stream of later values `Receiver<T>` or a combo representing a value now and many later `(T, Receiver<T>)`. There's a new sum type that encodes this concept `Effect<T>`. I'll have to get some examples up soon.

Also included is an RSX macro `dom!` for building DOM. It cuts down on the boiler plate quite a bit. After deciding to use conditional compilation to solve SSR the JSX macro became an orthogonal problem, but it was easy enough with [`syn-rsx`](https://crates.io/crates/syn-rsx):

```html
dom! {
    <button id="click_counter" on:click=tx.contra_map(...)>
        {("0 clicks", rx.branch_map(...))}
    </button>
}
```

**Comment by schell** [comment]
The `dom!` macro and initial SSR is now available on `master`. I'm going to work on hydration and then push a major version to crates.io.

**Comment by matthewharwood** [comment]
Heyhey maybe you're busy or my email missed ya. But you free on Thursday from 2:30~4?

**Comment by schell** [comment]
SSR is now available on `master`. I'm doing a documentation review before releasing to crates.io. Here's a summary of the big changes that were needed to make it happen:
* `Transmitter`s and `Receiver`s don't mutate to branch, forward, send, etc. 
* `Component::view` returns a `ViewBuilder` which builds up two closures - one that builds a view from scratch and one that hydrates a view. Alternatively you may create a `ViewBuilder` from a `View` (but you can't hydrate with it).
* greatly simplified the DOM building API
* uses RSX to cut down on boilerplate DOM building

**Comment by d4h0** [comment]
Thank you for your hard work, @schell, this is fantastic! :)

Does mogwai now also support code-splitting?

I'm thinking about creating a Rust/WASM-based static website generator similar to Gastby.js (which basically is a combination of a static website generator and a client-side JavaScript application framework).

One requirement for this is code-splitting (if a website has, say, 10.000 pages, it wouldn't be very performant if all 10k pages are loaded on the first visit... ;)).

(The great thing with such an approach is that you can include small (or big) apps and dynamic widgets, but otherwise you still have a static website, with all its benefits).

Also: It seems you didn't announce mogwai officially (at least on the sub-reddit).

What are your plans in this regard?

**Comment by schell** [comment]
> Does mogwai now also support code-splitting?

Code splitting is yet another feature I don't know much about, hehe! I'll have to read up on it. My naive idea is that if you have code that needs to be executed on the server _and_ in the browser then it should live in a library and be wrapped by an excutable that knows where it is to be executed. 

> It seems you didn't announce mogwai officially (at least on the sub-reddit)

Mogwai is still very much in flux with regard to its API. I've been reluctant to advertise for it during this time when breaking changes are happening often. Its capabilities are likely to only grow, but the specific functions and arguments are definitely still changing, and sometimes in large numbers! This ticket specifically changed the way DOM is built in a fundamental way and any application written with a previous version would require a pretty extensive update to bump versions. 

So `tl;dr` - the API is changing fast enough that I wouldn't feel comfortable telling many people to get onboard just yet, unless they're prepared for change.

**Comment by d4h0** [comment]
Thanks for answering my questions, @schell!

I'll post my answer to the code-splitting part in the [new issue](https://github.com/schell/mogwai/issues/41) I've created (I should have created a new issue from the beginning).

> So tl;dr - the API is changing fast enough that I wouldn't feel comfortable telling many people to get onboard just yet, unless they're prepared for change.

That makes sense. I'm actually in the same position with an unpublished crate I'm working on :)
